zebra: fix wrong linkdown flag

anlancs opened the pull request:
Add this one line command to one down interface (mask is 192.168.0.0/24):
```
ip link set dev enp2s0 up;ip route add 3.3.3.3/32 via 192.168.0.1 dev enp2s0 proto kernel onlink
```

Before:
```
K>* 3.3.3.3/32 [0/0] via 192.168.0.1, enp2s0 onlink linkdown, weight 1, 00:00:02
```

After:
```
K>* 3.3.3.3/32 [0/0] via 192.168.0.1, enp2s0 onlink, weight 1, 00:00:02
```

It's only display problem, just correct this linkdown flag.

ton31337 commented:
@Mergifyio backport stable/10.2 stable/10.1 stable/10.0 stable/9.1 stable/9.0

donaldsharp reviewed (commented):
Why is this not being set on link up?

anlancs commented:
> Why is this not being set on link up?

It is the time issue. If with two commands, there is no this issue.  Now it is combined into one command, so the UP message of netlink is slower. @donaldsharp

donaldsharp commented:
I am not seeing this problem:

```
sharpd@eva ~/f/t/t/bgp_bmp (master)> sudo vtysh -c "show ip route kernel"
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

K>* 0.0.0.0/0 [0/100] via 192.168.119.1, enp13s0, weight 1, 03:34:37
K>* 169.254.0.0/16 [0/1000] is directly connected, virbr2 linkdown, weight 1, 04:22:24
sharpd@eva ~/f/t/t/bgp_bmp (master)> sudo ip link set dev dummy1 up; sudo ip route add 3.3.3.3/32 via 192.168.0.1 dev dummy1 proto kernel onlink
sharpd@eva ~/f/t/t/bgp_bmp (master)> sudo vtysh -c "show ip route kernel"
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

K>* 0.0.0.0/0 [0/100] via 192.168.119.1, enp13s0, weight 1, 03:34:48
K>* 3.3.3.3/32 [0/0] via 192.168.0.1, dummy1 onlink, weight 1, 00:00:02
K>* 169.254.0.0/16 [0/1000] is directly connected, virbr2 linkdown, weight 1, 04:22:35
sharpd@eva ~/f/t/t/bgp_bmp (master)> 
```

How exactly do you reproduce?

anlancs commented:
@donaldsharp Use the REAL interface, try your enp13s0?

donaldsharp commented:
if the up comes later we should scan the kernel routes and set the flag there.  Hence this is the wrong spot to do the work.

anlancs commented:
```
root@debian# ip link set dev enp2s0 up;  ip route add 5.5.5.5/32 via 192.168.0.99 dev enp2s0  proto kernel
root@debian# vtysh -c 'show ip route kernel'
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, T - Table, v - VNC,
       V - VNC-Direct, D - SHARP, F - PBR, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

IPv4 unicast VRF default:
K>* 5.5.5.5/32 [0/0] via 192.168.0.99, enp2s0 linkdown, weight 1, 00:00:08
root@debian# ip addr show dev enp2s0
3: enp2s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 7c:d3:0a:14:95:43 brd ff:ff:ff:ff:ff:ff
    altname enx7cd30a149543
    inet 192.168.0.78/24 brd 192.168.0.255 scope global enp2s0
       valid_lft forever preferred_lft forever
    inet6 fe80::7ed3:aff:fe14:9543/64 scope link proto kernel_ll
       valid_lft forever preferred_lft forever
root@debian# vtysh -c 'show ip route kernel'
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, T - Table, v - VNC,
       V - VNC-Direct, D - SHARP, F - PBR, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

IPv4 unicast VRF default:
K>* 5.5.5.5/32 [0/0] via 192.168.0.99, enp2s0 linkdown, weight 1, 00:06:18
root@# ip route
5.5.5.5 via 192.168.0.99 dev enp2s0 proto kernel
```

It is very easy to get it wrong on my REAL machine.

anlancs committed:
zebra: fix wrong linkdown flag

When a route is received from the kernel while the associated interface
is in NO-CARRIER state, the kernel marks the nexthop with `RTNH_F_LINKDOWN`.
Zebra faithfully records this as `NEXTHOP_FLAG_LINKDOWN`.

Later, when the interface carrier comes up and `zebra_interface_nhg_reinstall()`
is called, the NHE is re-validated and re-installed into the kernel. However,
the stale `NEXTHOP_FLAG_LINKDOWN` was never cleared, causing "show ip route"
to incorrectly display "linkdown" for routes whose interface is fully operative.

Fix this by clearing `NEXTHOP_FLAG_LINKDOWN` in `zebra_nhg_set_valid_if_active()`
when the interface is confirmed operative.

Signed-off-by: anlan_cs <vic.lan@pica8.com>

anlancs commented:
I have updated this PR (commit log), can you review it again? Thanks  @donaldsharp @riw777
